### PR TITLE
#23: adds db creation logic

### LINKED ISSUE #23: Create and design database for stateful configuration
We will need a database access for many operations which require configuration, like commands that should make the bot do something in a special channel or let the bot send a specific message repeatedly. Things like the message or the channel for the specific command need to be persisted so the bot wont forget them once configured. 

This issue is only about creating a database with a fitting (and extendable) design for this and creating code to access it correctly.

### PR DESCRIPTION
**Pull request by SebastiantheAstorian**
Resolves #23 

Database tables and therefor their accessor method is subject to change according to command needs.

### CODE REVIEW

File: bot/__main__.py [hunk 1]
```diff
@@ -1,20 +1,21 @@
 """This is the main module of the SAM project."""
 
 from discord.ext import commands
+
```
Comment (PKlempe) [inline bot/__main__.py:4]: Why the newline between the two imports?

File: bot/persistence/database_connector.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+class DatabaseConnector():
+    """Class used to communicate with the database.
+     The database is created and initialized with, using the __init__ method
+     The other methods support getting or adding properties to the database.
+    """
+    _db_file: Union[str, None] = None
+
+    def __init__(self, db_file: str, init_script=None):
+        """ create a database connection to a SQLite database and create the default tables form the sql script in
```
Comment (PKlempe) [inline bot/persistence/database_connector.py:17]: Empty space in front of the word "create" which is also written in small caps.

File: bot/persistence/database_connector.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        init_db.sql
+
+        Args:
+            self: this object
```
Comment (PKlempe) [inline bot/persistence/database_connector.py:21]: `self` should never be mentionend in args.

File: bot/persistence/database_connector.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            db_file: The file name of the sqllite database file
```
Comment (PKlempe) [inline bot/persistence/database_connector.py:22]: Proper names and abbreviations always in upper case e.g. `SQLite` or `SQL`. Dot is missing at the end.

File: bot/persistence/database_connector.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            init_script: Optional sql script file name that will be run when method is called
```
Comment (PKlempe) [inline bot/persistence/database_connector.py:23]: `Filename` is one word not two and also abbrevation and missing dot again.

File: bot/persistence/database_connector.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+                    try:
+                        db_manager.execute(query)
+                    except Error as error:
+                        print("command could not be executed, skipping it: {0}".format(error))
+
+    def add_config_property(self, key: str, val: str):
+        """Adds a new configuration property to the database or updates an existing if the key already exists.
+
+        Args:
+            self: this object
```
Comment (PKlempe) [inline bot/persistence/database_connector.py:40]: As before: Never mention `self` in args.

File: bot/persistence/__init__.py [hunk 7]
```diff
@@ -0,0 +1,5 @@
+"""Module containing persistence logic and accessors for Database. The only needed class to initialize and
+access the database is the DatabaseConnector.
```
Comment (PKlempe) [inline bot/persistence/__init__.py:2]: Docstring format isn't 100% accurate. Take a look at this:
[https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)

File: bot/constants.py [hunk 8]
```diff
@@ -13,3 +14,12 @@
 DISCORD_BOOST_LVL3_CAP = 30
 
 EMBED_INFO_COLOR = 0x1E90FF
+
+# Database
+DB_FILE_PATH = os.getenv("DB_FILE_PATH")
+#absolute path to the init script
+DB_INIT_SCRIPT = os.path.join(os.path.dirname(os.path.abspath(__file__)), "persistence/resources/init_db.sql")
+
+# Database Queries
```
Comment (PKlempe) [inline bot/constants.py:23]: I'm not fully convinced that we will need **the exact same** queries multiple times in our code and I'm therfore unsure if we should store them as constants.

File: bot/constants.py [hunk 9]
```diff
@@ -2,9 +2,10 @@
 
 import os
 from dotenv import load_dotenv, find_dotenv
+
 load_dotenv(find_dotenv())
 
-DISCORD_BOT_TOKEN = os.getenv("DISCORD_BOT_TOKEN")
+DISCORD_API_KEY = os.getenv("DISCORD_API_KEY")
```
[the remark this answers is not in the document]
Comment (SebastiantheAstorian) [reply bot/constants.py:8]: You named it like that in the .env example file. The code didnt work so I had to rename either this or the other.

File: bot/constants.py (hunk 8 above)
Comment (SebastiantheAstorian) [reply bot/constants.py:23]: It makes it easier to manage and IMO to read. I will use a dedicated query constants file in the persistence module

File: bot/constants.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (PKlempe) [reply bot/constants.py:8]: Oh, than I must have forgot to edit the example file. Please use `DISCORD_BOT_TOKEN`.

File: bot/persistence/__init__.py (hunk 7 above)
Comment (SebastiantheAstorian) [reply bot/persistence/__init__.py:2]: I don't believe I am thaaat far away from the recommended google style, as it allows for a lot of freedom, however I will still adapt it as good as possible

File: bot/persistence/database_connector.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    """Class used to communicate with the database.
+
+    The database is created and initialized with, using the __init__ method
+    The other methods support getting or adding properties to the database.
+    """
+
+    _db_file: Union[str, None] = None
+
+    def __init__(self, db_file: str, init_script=None):
+        """Create a database connection to a SQLite database and create the default tables form the SQL script in
```
Comment (PKlempe) [inline bot/persistence/database_connector.py:23]: How about:
> [...] and create the tables specified in the SQL script called "init_db.sql".

File: bot/persistence/database_connector.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        init_db.sql.
+
+        Args:
+            db_file (str): The filename of the SQLite database file.
+            init_script (Union[str, None]): Optional SQL script filename that
```
Comment (PKlempe) [inline bot/persistence/database_connector.py:28]: You should check your configured line width. Seems a bit short if you make linebreaks this early. Mine is 120.

File: bot/persistence/database_connector.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+                will be run when the method is called.
+        """
+        self._db_file = db_file
+        with DatabaseManager(self._db_file) as db_manager:
```
Comment (PKlempe) [inline bot/persistence/database_connector.py:32]: Out of curiosity: When does the `with` statement release resources? After it's done with its code or when our DatabaseConnector is destructed?

File: bot/persistence/database_connector.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+            if init_script is not None:
+                queries = self.parse_sql_file(init_script)
+                for query in queries:
+                    try:
+                        db_manager.execute(query)
+                    except Error as error:
+                        print("command could not be executed, skipping it: {0}".format(error))
```
Comment (PKlempe) [inline bot/persistence/database_connector.py:39]: First word should start with capital letter.

File: bot/persistence/database_connector.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+
+    def add_config_property(self, key: str, val: str):
+        """Adds a new configuration property to the database or updates an
+        existing if the key already exists.
```
Comment (PKlempe) [inline bot/persistence/database_connector.py:43]: ... an existing one ...

File: bot/persistence/database_connector.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+            if init_script is not None:
+                queries = self.parse_sql_file(init_script)
+                for query in queries:
+                    try:
+                        db_manager.execute(query)
+                    except Error as error:
+                        print("command could not be executed, skipping it: {0}".format(error))
+
+    def add_config_property(self, key: str, val: str):
+        """Adds a new configuration property to the database or updates an
```
Comment (PKlempe) [inline bot/persistence/database_connector.py:42]: Configuration property? Do you mean an attribute?

File: bot/persistence/database_manager.py [hunk 16]
```diff
@@ -0,0 +1,57 @@
+"""Context manager for managing database connections"""
```
Comment (PKlempe) [inline bot/persistence/database_manager.py:1]: Missing dot.

File: bot/persistence/database_connector.py [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+            if row is not None:
+                return row[0]
+            return None
+
+    @staticmethod
+    def parse_sql_file(filename: str) -> List[str]:
+        """Parses a SQL script to read all queries/commands it contains.
+
+        Args:
+            filename (str): the filename of the init file. Can also be a path.
```
Comment (PKlempe) [inline bot/persistence/database_connector.py:87]: *The

File: bot/persistence/database_connector.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+            db_manager.commit()
+
+    def get_property(self, key: str):
+        """Searches the config table of the db for a specific property.
+
+        Args:
+            key (str): The property key to search for.
+
+        Returns:
+            Union[str, None]: the property value for the specified key or None
```
Comment (PKlempe) [inline bot/persistence/database_connector.py:68]: *The

File: bot/persistence/database_connector.py [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+
+    @staticmethod
+    def parse_sql_file(filename: str) -> List[str]:
+        """Parses a SQL script to read all queries/commands it contains.
+
+        Args:
+            filename (str): the filename of the init file. Can also be a path.
+
+        Returns:
+            List[str]: A list of strings with each entry being a SQL-query.
```
Comment (PKlempe) [inline bot/persistence/database_connector.py:90]: Hyphen not needed in English.

File: bot/persistence/database_manager.py [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+import sqlite3
+from sqlite3 import Error
+
+
+class DatabaseManager:
+    """Context Manager class allowing for simple and resilient accessing of the
+    db.
+
+    Example:
+        This class can be used as followed::
```
Comment (PKlempe) [inline bot/persistence/database_manager.py:12]: Double colon.

File: bot/persistence/database_manager.py [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+        After the last line in the indented block is executed, the connection
+        will be automatically closed via the exit method.
+    """
+    connection = None
+
+    def __init__(self, db_file: str):
+        """Initializes the context manager with the db file name.
+
+        Args:
+            db_file (str): the name (or path) to the db file
```
Comment (PKlempe) [inline bot/persistence/database_manager.py:27]: `*The` and missing dot.

File: bot/persistence/database_manager.py [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+        """
+        self.db_file = db_file
+        self.connection = None
+
+    def __enter__(self):
+        """Entry method of the context manager.
+
+        Opens the db connection using the sqllite3 library.
```
Comment (PKlempe) [inline bot/persistence/database_manager.py:35]: *sqlite3

File: bot/persistence/database_manager.py [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+        try:
+            self.connection = sqlite3.connect(self.db_file)
+        except Error as error:
+            print(error)
+        return self.connection
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Exit method of the context manager.
+
+        Closes the database connection after the code has executed.
```
Comment (PKlempe) [inline bot/persistence/database_manager.py:49]: *has been executed.

File: bot/persistence/queries.py [hunk 24]
```diff
@@ -0,0 +1,5 @@
+"""This module contains template string constants to be used for database queries and commands
```
Comment (PKlempe) [inline bot/persistence/queries.py:1]: Missing dot.

File: bot/test/test_db.py [hunk 25]
```diff
@@ -0,0 +1,20 @@
+"""Tests for the database and other persistence classes"""
```
Comment (PKlempe) [inline bot/test/test_db.py:1]: Missing dot.

File: bot/test/test_db.py [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
+import os
+from bot import constants
+from bot.persistence import DatabaseConnector
+
+
+def test_db():
+    """Tests if inserts and reads from the database work. Initializes the
+    database, adds a single value, queries the same value by key and finally
+    deletes the db file. Passes if the returned value is found (ie. not None)
+    and equal to the original one
```
Comment (PKlempe) [inline bot/test/test_db.py:12]: Missing dot.

File: bot/util/utilities.py [hunk 27]
```diff
@@ -10,7 +10,11 @@ class UtilitiesCog(commands.Cog):
     """Cog for Utility Functions."""
 
     def __init__(self, bot):
-        """Initializes the Cog."""
+        """Initializes the Cog.
+
+        Args:
+            bot:
```
Comment (PKlempe) [inline bot/util/utilities.py:16]: Missing type and description.

File: bot/util/utilities.py [hunk 28]
```diff
@@ -47,7 +53,7 @@ async def server_info(self, ctx):
         embed.add_field(name="Server Boost <:server_boost:730390579699122256>", value=embed_strings[1], inline=True)
         embed.add_field(name="Server Features :tools:", value=embed_strings[2], inline=True)
         embed.add_field(name="Mitglieder :man_raising_hand:", value=embed_strings[3], inline=True)
-        embed.add_field(name="Kanäle :dividers:", value=embed_strings[4], inline=True)
+        embed.add_field(name="KanÃƒÆ’Ã†â€™Ãƒâ€šÃ‚Â¤le :dividers:", value=embed_strings[4], inline=True)
```
Comment (PKlempe) [inline bot/util/utilities.py:56]: I don't know what happend here...

File: bot/util/utilities.py [hunk 29]
```diff
@@ -148,12 +159,12 @@ def generate_features_string(features):
         "DISCOVERABLE":             "In Server-Browser",
         "FEATURABLE":               "Featurable",
         "COMMERCE":                 "Commerce",
-        "PUBLIC":                   "Öffentlich",
-        "NEWS":                     "News-Kanäle",
+        "PUBLIC":                   "ÃƒÆ’Ã†â€™ÃƒÂ¢Ã¢â€šÂ¬Ã¢â‚¬Å“ffentlich",
```
Comment (PKlempe) [inline bot/util/utilities.py:162]: Same problem as mentioned before.

**Pull request merged**